feat: expose workspace statuses (with details) as a prometheus metric

dannykopping opened the pull request:
Implements https://github.com/coder/coder/issues/12462

This produces the following metric:

```promql
coderd_api_workspace_detail{status="succeeded",template_name="docker",template_version="distracted_hopper6",workspace_name="test_workspace",workspace_owner="danny",workspace_transition="start"}
```

**Notes to reviewers**:

- Minor behaviour change: previously if a metric existed but then subsequent queries loaded no entries, it would no reset the gauge. Additionally, we set the state of the gauge on startup to help with testing. Both of these would create a weird user-experience in my view so I got rid of them, but it has some small side-effects which I call out in the code
- I've reduced the refresh rate of `ActiveUsers` and `Workspaces` metrics to a minute; 5 minutes seemed quite excessive and may not allow operators to act fast enough; not sure if there was a reason these were different to the other metrics in this file
  - we should actually move towards the recommended _"collector"_ pattern since this is acting like an exporter: https://prometheus.io/docs/instrumenting/writing_exporters/#collectors; in other words, only run the queries when the metrics are collected (by a request to `/metrics`, for example)
  - speaking of best practices, I created https://github.com/coder/coder/issues/12744 to deprecate & rename metrics which are gauges but use `_total` suffixes which are generally reserved for counters
- I've also made the `Workspace` tests run with a slower interval (25ms instead of 1ms) and a faster cut-off (1s vs 10s)

dannykopping committed:
Add metric to show operators what statuses workspaces are in, with relevant detail

Light refactoring

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Clear gauges if no db entries are found

Light refactoring

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Change default refresh rates of Agent and Workspaces calls which seemed inappropriately slow

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Added tests for workspace detail metrics

Moderate refactoring

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Refactor common funcs

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Remove unnecessary metric from test

Use stronger rand

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping commented on cli/server_test.go:
```diff
@@ -973,26 +973,20 @@ func TestServer(t *testing.T) {
 					hasActiveUsers = true
 					continue
 				}
-				if strings.HasPrefix(scanner.Text(), "coderd_api_workspace_latest_build_total") {
-					hasWorkspaces = true
-					continue
-				}
```
Side-effect of clearing the gauge when no db rows are loaded.
This wasn't strictly necessary for the test since other manually-registered metrics are included to validate this test.

dannykopping committed:
Use NoLock variety to fix hung tests

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Fix duration on windows builds

Signed-off-by: Danny Kopping <danny@coder.com>

johnstcn commented on coderd/database/queries.sql.go:
```diff
@@ -12499,30 +12501,31 @@ type GetWorkspacesParams struct {
+	LatestBuildCanceledAt  sql.NullTime             `db:"latest_build_canceled_at" json:"latest_build_canceled_at"`
+	LatestBuildError       sql.NullString           `db:"latest_build_error" json:"latest_build_error"`
+	LatestBuildTransition  WorkspaceTransition      `db:"latest_build_transition" json:"latest_build_transition"`
+	LatestBuildStatus      NullProvisionerJobStatus `db:"latest_build_status" json:"latest_build_status"`
```
Is there a way to make this be non-nullable in the query?

dannykopping replied:
Afraid not, looks like the `latest_build` attributes are the result of a left join: https://github.com/coder/coder/blob/843d65035f0c1c436e24ef2ac5121b7c5b58784d/coderd/database/queries.sql.go#L12172-L12218

dannykopping replied:
I suppose by definition a workspace cannot exist without a build...

I see in `workspace_builds`:

```sql
create table if not exists public.workspace_builds
(
    id                  uuid                                                                                not null
        primary key,
    created_at          timestamp with time zone                                                            not null,
    updated_at          timestamp with time zone                                                            not null,
    workspace_id        uuid                                                                                not null
        references public.workspaces
            on delete cascade,
...
```

The workspace foreign key is not null, so I think this left join may be inappropriate?

dannykopping replied:
Hhmm. I changed it and ran `make gen` but it's still being defined as a nullable field.
Also, one of its sibling fields `latest_build_transition` is not nullable and is derived from the same join, so actually all of the above about the join was ostensibly was invalid.

I'll need to dig a little further.

dannykopping replied:
Ah! I was basically on the correct track!

`provisioner_jobs` was being left joined as well, and since that also has a non-nullable foreign key with `workspace_builds.job_id`, it is safe to use an inner join. I can actually leave the outer lateral join as it is but just change this to correct the generated code.

dannykopping replied:
@johnstcn see [b118044](https://github.com/coder/coder/pull/12762/commits/b118044549d20a76ffe6c1a45a440a64e3f7f9a9)

johnstcn commented on coderd/database/queries.sql.go:
```diff
@@ -12405,7 +12406,7 @@ WHERE
 ), filtered_workspaces_order AS (
 	SELECT
-		fw.id, fw.created_at, fw.updated_at, fw.owner_id, fw.organization_id, fw.template_id, fw.deleted, fw.name, fw.autostart_schedule, fw.ttl, fw.last_used_at, fw.dormant_at, fw.deleting_at, fw.automatic_updates, fw.favorite, fw.template_name, fw.template_version_id, fw.template_version_name, fw.username, fw.latest_build_completed_at, fw.latest_build_canceled_at, fw.latest_build_error, fw.latest_bui ...
+		fw.id, fw.created_at, fw.updated_at, fw.owner_id, fw.organization_id, fw.template_id, fw.deleted, fw.name, fw.autostart_schedule, fw.ttl, fw.last_used_at, fw.dormant_at, fw.deleting_at, fw.automatic_updates, fw.favorite, fw.template_name, fw.template_version_id, fw.template_version_name, fw.username, fw.latest_build_completed_at, fw.latest_build_canceled_at, fw.latest_build_error, fw.latest_bui ...
```
follow-up idea: it could be useful to have this field in the `codersdk.Workspace` as well :-)

dannykopping replied:
@johnstcn and I spoke offline to elaborate on this, apparently we convert between the `database.Workspace` and `codersdk.Workspace` types in some situations (`convertWorkspace`, for example), and it would be good to maintain parity between them. This would be an enhancement for a subsequent PR, we agreed.

dannykopping committed:
Changing join type so that latest_build_status is non-nullable

Signed-off-by: Danny Kopping <danny@coder.com>

johnstcn reviewed (approved):
Smoke-tested:

After creating workspace:

```
curl -fsSL http://localhost:21112 | grep coderd_api_workspace_detail
# HELP coderd_api_workspace_detail The current workspace details by template, transition, owner, and status.
# TYPE coderd_api_workspace_detail gauge
coderd_api_workspace_detail{status="succeeded",template_name="docker",template_version="nice_hofstadter1",workspace_name="test",workspace_owner="admin",workspace_transition="start"} 1
```

After stopping a workspace and waiting approx 1 minute:

```
curl -fsSL http://localhost:21112 | grep coderd_api_workspace_detail
# HELP coderd_api_workspace_detail The current workspace details by template, transition, owner, and status.
# TYPE coderd_api_workspace_detail gauge
coderd_api_workspace_detail{status="succeeded",template_name="docker",template_version="nice_hofstadter1",workspace_name="test",workspace_owner="admin",workspace_transition="stop"} 1
```

dannykopping commented:
Based on @johnstcn's comment and a little time away from this, I'm thinking the current metric name is not quite explanatory enough.

I'm thinking of `coderd_api_workspace_latest_build_status`. Thoughts?

mtojek commented on coderd/prometheusmetrics/prometheusmetrics.go:
```diff
@@ -72,36 +74,43 @@ func ActiveUsers(ctx context.Context, registerer prometheus.Registerer, db datab
+	workspaceDetails := prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "coderd",
+		Subsystem: "api",
+		Name:      "workspace_detail",
```
Naming: can we just call it `coder_workspaces`? We have `coderd_agents_apps`, `coderd_insights_parameters`. I wouldn't pair with `api` specifically.

EDIT:

Following the Prometheus naming convention (credits to @spikecurtis for linking in a review a while ago), we could also consider `coder_workspace_info`.

dannykopping replied:
Regarding the `api` subsystem, I just copied how the other metric defined in this function is named. I agree that bit is probably extraneous and am happy to take it out.

`_info` metrics are usually reserved for "static" metrics, such as `build_info` and things like that.

mtojek replied:
> Regarding the api subsystem, I just copied how the other metric defined in this function is named. I agree that bit is probably extraneous and am happy to take it out.

👍 

> _info metrics are usually reserved for "static" metrics, such as build_info and things like that.

TIL! Thanks for the explanation.

mtojek commented on coderd/prometheusmetrics/prometheusmetrics.go:
```diff
@@ -72,36 +74,43 @@ func ActiveUsers(ctx context.Context, registerer prometheus.Registerer, db datab
+		return nil, err
+	}
+
+	workspaceDetails := prometheus.NewGaugeVec(prometheus.GaugeOpts{
```
Thinking loud:

During the last scale tests we hit 2000 workspaces, does it mean that we're going to swamp our Prometheus endpoint with details for all of them? If so, maybe we should make it configurable? cc @johnstcn

johnstcn replied:
See also Spike's comment: https://github.com/coder/coder/issues/12462#issuecomment-1985129901

dannykopping replied:
Oh wow, it seems I misunderstood the initial ask.

I was under the incorrect assumption that @bpmct wanted a metric **to match** the format described in the issue, and "the builds metric" which he was referring to was the existing `coderd_api_workspace_latest_build_total` metric 🤦

`coderd_workspace_builds_total` already exists!

```promql
# HELP coderd_workspace_builds_total The number of workspaces started, updated, or deleted.
# TYPE coderd_workspace_builds_total counter
coderd_workspace_builds_total{status="failed",template_name="docker",template_version="vibrant_davinci2",workspace_name="asfsasfafs",workspace_owner="danny",workspace_transition="START"} 1
coderd_workspace_builds_total{status="success",template_name="docker",template_version="beautiful_visvesvaraya1",workspace_name="2",workspace_owner="danny",workspace_transition="START"} 1
coderd_workspace_builds_total{status="success",template_name="docker",template_version="beautiful_visvesvaraya1",workspace_name="asfsasfafs",workspace_owner="danny",workspace_transition="START"} 1
coderd_workspace_builds_total{status="success",template_name="docker",template_version="beautiful_visvesvaraya1",workspace_name="asfsasfafs",workspace_owner="danny",workspace_transition="STOP"} 1
coderd_workspace_builds_total{status="success",template_name="docker",template_version="mystifying_driscoll3",workspace_name="2",workspace_owner="danny",workspace_transition="STOP"} 1
```

Plus we don't have to worry about cardinality because these metrics are written by the provisioner when executing a job, not reading from the database state.

I have amended the new metric to drop the `workspace` label.

I _think_ this will satisfy the requirements but @bpmct please confirm.

```promql
# HELP coderd_workspace_latest_build_status The current workspace statuses by template, transition, and owner.
# TYPE coderd_workspace_latest_build_status gauge
coderd_workspace_latest_build_status{status="failed",template_name="docker",template_version="happy_feynman7",workspace_owner="danny",workspace_transition="stop"} 1
coderd_workspace_latest_build_status{status="succeeded",template_name="docker",template_version="beautiful_visvesvaraya1",workspace_owner="danny",workspace_transition="start"
} 2
coderd_workspace_latest_build_status{status="succeeded",template_name="docker",template_version="distracted_hopper6",workspace_owner="danny",workspace_transition="start"} 1
coderd_workspace_latest_build_status{status="succeeded",template_name="docker",template_version="ecstatic_greider3",workspace_owner="danny",workspace_transition="start"} 1
```

bpmct replied:
> I _think_ this will satisfy the requirements but @bpmct please confirm.
> 
> ```
> # HELP coderd_workspace_latest_build_status The current workspace statuses by template, transition, and owner.
> # TYPE coderd_workspace_latest_build_status gauge
> coderd_workspace_latest_build_status{status="failed",template_name="docker",template_version="happy_feynman7",workspace_owner="danny",workspace_transition="stop"} 1
> coderd_workspace_latest_build_status{status="succeeded",template_name="docker",template_version="beautiful_visvesvaraya1",workspace_owner="danny",workspace_transition="start"
> } 2
> coderd_workspace_latest_build_status{status="succeeded",template_name="docker",template_version="distracted_hopper6",workspace_owner="danny",workspace_transition="start"} 1
> coderd_workspace_latest_build_status{status="succeeded",template_name="docker",template_version="ecstatic_greider3",workspace_owner="danny",workspace_transition="start"} 1
> ```

Yep!

mtojek reviewed (commented):
High level review round

mtojek commented:
> I'm thinking of coderd_api_workspace_latest_build_status. Thoughts?

I wonder if it isn't too late for renaming as people may use it on their dashboards.

dannykopping commented:
> > I'm thinking of coderd_api_workspace_latest_build_status. Thoughts?
> 
> I wonder if it isn't too late for renaming as people may use it on their dashboards.

This is a new metric being added by this PR.

mtojek commented:
My bad, I assumed that this is `coderd_workspace_builds_total` being refactored.

mafredri commented on coderd/database/queries/workspaces.sql:
```diff
@@ -118,7 +119,7 @@ LEFT JOIN LATERAL (
 	FROM
 		workspace_builds
-	LEFT JOIN
+	JOIN
```
I would avoid making this change. I believe there's a window of time when there exists a new build but it's not assigned to a job. So the nullability is something that should be handled.

This logic change essentially results in returning the previous build until the job has started.

dannykopping replied:
I'm not sure if that's possible?

```sql
create table if not exists workspace_builds
(
    id                  uuid                                                                                not null
    ...
    workspace_id        uuid                                                                                not null
        references public.workspaces
            on delete cascade,
    ...
    job_id              uuid                                                                                not null
        unique
        references public.provisioner_jobs
            on delete cascade,
    ...
);
```

These keys are all non-nullable.
Is there something I'm missing?

dannykopping replied:
I should say though that I'd rather not make this change if you in any way suspect this will cause issues.
The way I had it prior to https://github.com/coder/coder/pull/12762#discussion_r1539514142 worked fine.

spikecurtis replied:
Builds are inserted with their respective job in a single transaction.  c.f. `wsbuilder`

Also, Danny's right that job_id is `NOT NULL` with a foreign key constraint, so even if we broke `wsbuilder` it still wouldn't be possible to have a build with no job.

mafredri replied:
Given the constraint, the change is perfectly fine 👍. I should’ve checked it myself before commenting. Sorry for the noise @dannykopping.

mafredri replied:
The DB schema should probably be updated to include NOT NULL on the foreign keys, though. Otherwise NULL entries are not enforced by the DB.

dannykopping replied:
No worries @mafredri!

I just generated the DDL from an existing database in my IDE (GoLand). The above also has `not null` but you have to scroll over a bit to the right.

mafredri replied:
Haha, it was perfectly hidden by the code block size 😂.

dannykopping commented on testutil/duration.go:
```diff
@@ -9,10 +9,11 @@ import (
-	WaitMedium    = 15 * time.Second
-	WaitLong      = 25 * time.Second
-	WaitSuperLong = 60 * time.Second
+	WaitSuperShort = time.Second
```
You're right; I added this in because some log lines were spewing too much during a test (I added logging when no rows are found), but it'll be safer to just set an appropriate log level.

10 secs felt a bit too long as well, but I guess we have this in a bunch of places so it's fine.

dannykopping committed:
Remove SuperShort duration as extraneous

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Remove workspace cardinality; rename for clarity

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping committed:
Fixing tests

Signed-off-by: Danny Kopping <danny@coder.com>

spikecurtis commented on coderd/prometheusmetrics/prometheusmetrics.go:
```diff
@@ -110,16 +118,53 @@ func Workspaces(ctx context.Context, registerer prometheus.Registerer, db databa
+			return
+		}
+
+		workspaceStatuses.Reset()
```
this has a race condition where if the registry collects the metrics before the for-loop below is complete, it will get partial data.  This can result in glitchy stats were they are zeroed out for a single data point and then return to "normal."  If anyone puts alerts on those stats it could result in false alerting in addition to visual noise on the dashboard.

I realize this is following the pattern of the existing metric, which has the same bug.

Instead these stats should be a custom collector that queries the database and returns the requested data on demand.

dannykopping replied:
> Instead these stats should be a custom collector that queries the database and returns the requested data on demand.
Definitely agree! Called this out in the description as well.

I'm not sure if we can fix this race without the change to the collector pattern. Do you have any ideas?
In reality the race would have to be on the nanosecond level because this loop will iterate very quickly, but your point is absolutely valid.

I'll add a follow-up issue for this in any case.

dannykopping replied:
https://github.com/coder/coder/issues/12792

dannykopping committed:
Merge branch 'main' of github.com:/coder/coder into dk/workspace-metric

mtojek commented on coderd/prometheusmetrics/prometheusmetrics.go:
```diff
@@ -110,16 +118,53 @@ func Workspaces(ctx context.Context, registerer prometheus.Registerer, db databa
+				workspaceStatuses.Reset()
+			}
+
+			logger.Warn(ctx, "failed to load active workspaces", slog.Error(err))
```
nit: maybe `logger.Error`?

dannykopping replied:
I wouldn't strictly call it an error because it could be an instance of `sql.ErrNoRows` which is valid but undesirable. To be the most correct I would make it warn in case of `sql.ErrNoRows` and error in all other cases, but that feels unnecessary.

dannykopping replied:
Believe so, yes.

mtojek reviewed (approved):
Looks like it works as intended 👍 I left some comments on naming, but definitely not blockers for pushing.

dannykopping committed:
Merge branch 'main' of github.com:/coder/coder into dk/workspace-metric

dannykopping committed:
Renaming for clarity

Signed-off-by: Danny Kopping <danny@coder.com>

dannykopping commented on coderd/prometheusmetrics/prometheusmetrics.go:
```diff
@@ -72,36 +74,42 @@ func ActiveUsers(ctx context.Context, registerer prometheus.Registerer, db datab
 	}
 
-	gauge := prometheus.NewGaugeVec(prometheus.GaugeOpts{
+	workspaceStatusTotals := prometheus.NewGaugeVec(prometheus.GaugeOpts{
```
That's fair; I've updated accordingly: 8e6cde9e7

dannykopping commented:
I believe I've addressed all the comments. I'll merge, but if there any subsequent thoughts, I'll amend those in a follow-up.
Thanks for the reviews!
